stor/dx.py

```diff
@@ -114,95 +163,704 @@ def temp_url(self, lifetime=300, filename=None):
+                results = p.download_objects('dest/folder', ['subdir/f1.txt',
+                                                             'subdir/f2.txt'])
+                print results
+                {
+                    'subdir/f1.txt': 'dest/folder/subdir/f1.txt',
+                    'subdir/f2.txt': 'dest/folder/subdir/f2.txt'
+                }
+
+            To download full dx paths relative to a download path::
+
+                from stor import Path
+                p = Path('dx://project:/dir/')
+                results = p.download_objects('dest/folder', [
+                    'dx://project:/dir/subdir/f1.txt',
+                    'dx://project:/dir/subdir/f2.txt'
+                ])
+                print results
+                {
+                    'dx://project:/dir/subdir/f1.txt': 'dest/folder/subdir/f1.txt',
+                    'dx://project:/dir/subdir/f2.txt': 'dest/folder/subdir/f2.txt'
+                }
+        """
+        def dx_parent_dir(dx_obj1, dx_obj2):
+            """Checks if dx_obj2 is a sub-path of dx_obj1"""
+            if not dx_obj1.resource:
+                return dx_obj2.resource and dx_obj2.project == dx_obj1.project
+            return dx_obj2.startswith(utils.with_trailing_slash(dx_obj1))
+        source = self
+        if source == (self.drive+self.project):  # need to convert dx://proj to dx://proj:
+            source = DXPath(self + ':')
+
+        for obj in objects:
+            if utils.is_dx_path(obj) and not dx_parent_dir(source, DXPath(obj)):
+                    raise ValueError(
+                        '"%s" must be child of download path "%s"' % (obj, self))
+        # Convert requested download objects to full object paths
+        objs_to_download = {
+            obj: DXPath(obj) if utils.is_dx_path(obj) else source / obj
+            for obj in objects
+        }
+        results = {}
+        for obj, dx_obj in objs_to_download.items():
+            dest_resource = dx_obj[len(source):].lstrip('/')
+            dest_obj = PosixPath(dest) / dest_resource
+            dx_obj.copy(dest_obj)
+            results[obj] = dest_obj
+        return results
+
+    def download(self, dest, **kwargs):
+        """Download a directory.
+
+        Args:
+            dest (Path): The output directory
+
+        Raises:
+            ValueError: When source or dest path is not a directory
+        """
+        if not self.isdir():
+            raise NotFoundError('No folder or project was found at given path ({}) on DNAnexus'
+                                .format(self))
+        with _propagate_dx_exceptions():
+            dxpy.download_folder(
+                project=self.canonical_project,
+                destdir=dest,
+                folder=self.resource or '/'
+            )
+
+    def upload(self, to_upload, **kwargs):
+        """Upload a list of files and directories to a directory.
+
+        Note that unlike swift, this is not a batch level operation.
```

Comment:
I don't think this is true for swift, each API operation is separate (just swiftclient has a batch API for it). just remove